service/topology_coordinator.cc

```diff
@@ -1634,10 +1651,31 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
                         auto replaced_id = std::get<replace_param>(node.req_param.value()).replaced_id;
                         auto it = _topo_sm._topology.normal_nodes.find(replaced_id);
                         SCYLLA_ASSERT(it != _topo_sm._topology.normal_nodes.end());
-                        SCYLLA_ASSERT(!it->second.ring.tokens.empty() && it->second.state == node_state::normal);
+                        SCYLLA_ASSERT(it->second.state == node_state::normal);
 
                         topology_mutation_builder builder(node.guard.write_timestamp());
 
+                        // If a zero-token node is replacing another zero-token node,
+                        // we can instantly move the new node's state to normal.
+                        if (it->second.ring.tokens.empty()) {
+                            node = retake_node(co_await remove_from_group0(std::move(node.guard), replaced_id), node.id);
```

Comment:
What if retake fail here?

Reply:
If we want to keep replaced node a member of group 0, we could only make it non-voter here, and remove it completely in `left` state.

There is a similar problem in the replace of token-owning node scenario: in `transition_state::write_both_read_new`, `node_state::replacing`, we first remove from group 0, then move it to `left`. It's possible we fail in the middle. However it's a bit different case because `write_both_read_new` means no rollback, the coordinator will retry until it succeeds, so eventually the node will be moved to `left`. Also, I'm not sure, but isn't the replaced node already banned at this point?

I know that we ban at the beginning of removenode.
Do we also ban at the beginning of replace? If we do, then removing from group 0 here should be fine, even if we fail the replace operation.

Reply:
We will catch it in `topology_coordinator::handle_topology_coordinator_error` and the topology coordinator will retry.

Maybe we should roll back the join operation instead. When do we want to roll back and when we don't? It's not clear to me. In the transition handler, there are a lot of cases when we don't roll back the operation. For example, there are similar `retake_node(co_await remove_from_group0(` calls.

Reply:
> However it's a bit different case because `write_both_read_new` means no rollback, the coordinator will retry until it succeeds, so eventually the node will be moved to `left`.

Rollback also won't happen here because we don't catch the exception and don't set `_rollback`. The exception will be caught in `handle_topology_coordinator_error` and the topology coordinator will continue retrying until it succeeds.

Maybe we should roll back here instead?

> I know that we ban at the beginning of removenode. Do we also ban at the beginning of replace? If we do, then removing from group 0 here should be fine, even if we fail the replace operation.

It looks like we ban all ignored nodes:

```cpp
future<> storage_service::topology_state_load(state_change_hint hint) {
    ...
    for (auto& id : _topology_state_machine._topology.ignored_nodes) {
        // Ban all ignored nodes. We do not allow them to go back online
        co_await _messaging.local().ban_host(locator::host_id{id.uuid()});
    }
```

And it includes the replaced node (the code below is called in the join request handler):

```cpp
std::unordered_set<raft::server_id> storage_service::ignored_nodes_from_join_params(const join_node_request_params& params) {
    std::unordered_set<raft::server_id> ignored_nodes;

    if (!params.ignore_nodes.empty()) {
        std::list<locator::host_id_or_endpoint> ignore_nodes_params;
        for (const auto& n : params.ignore_nodes) {
            ignore_nodes_params.emplace_back(n);
        }

        ignored_nodes = find_raft_nodes_from_hoeps(ignore_nodes_params);
    }

    if (params.replaced_id) {
        // insert node that should be replaced to ignore list so that other topology operations
        // can ignore it
        ignored_nodes.insert(*params.replaced_id);
    }

    return ignored_nodes;
}
```

Reply:
> Rollback also won't happen here because we don't catch the exception and don't set _rollback. The exception will be caught in handle_topology_coordinator_error and the topology coordinator will continue retrying until it succeeds.

> Maybe we should roll back here instead?

Depends: if the replacing node dies, will we be able to progress here? If not, we should rollback. Otherwise the coordinator would get forever stuck.

Reply:
I don't see a reason why we wouldn't be able to finish the replace operation if the replacing node dies. `remove_from_group0` and `update_topology_state` don't require the replacing  node to be alive.

Reply:
I did a manual test that you asked for and it passed with no problems.